File: backend/app/schemas/food_log.py
"""
FoodLog request/response schemas.

Key design decision: The client always sends raw_input (what user typed).
The server parses and estimates — never trust client-provided calorie values
unless estimation_source = 'manual' (user explicitly entered exact values).

Three logging modes:
1. Quick log: raw_input only → server parses + estimates
2. Manual log: full nutrition values → stored as-is with source='manual'
3. Corrected log: update existing log with user's verified values
"""
from datetime import datetime
from uuid import UUID

from pydantic import Field, field_validator, model_validator

from app.schemas.common import (
    ConfidenceLevel,
    EstimationSource,
    MealType,
    TrackBaseSchema,
)


# ── Nested Schemas ─────────────────────────────────────────────────────────────

class NutritionData(TrackBaseSchema):
    """Macro and micro nutrition values."""
    calories: float = Field(ge=0.0, le=10000.0)
    protein_g: float | None = Field(default=None, ge=0.0, le=1000.0)
    carbs_g: float | None = Field(default=None, ge=0.0, le=2000.0)
    fat_g: float | None = Field(default=None, ge=0.0, le=500.0)
    fiber_g: float | None = Field(default=None, ge=0.0, le=200.0)


class PortionData(TrackBaseSchema):
    """Portion size in human-readable and metric forms."""
    description: str | None = Field(default=None, max_length=255,
                                     description="e.g. 'medium bowl', '2 pieces'")
    grams: float | None = Field(default=None, ge=0.0, le=10000.0)


# ── Request Schemas ────────────────────────────────────────────────────────────

class FoodLogCreate(TrackBaseSchema):
    """
    Body for POST /api/v1/food-logs/

    Two valid modes:
    1. Quick: provide raw_input only. Server estimates nutrition via AI pipeline.
    2. Manual: provide food_name + calories (+ optionally macros). Stored directly.
    """
    # Quick log mode: user types natural language
    raw_input: str | None = Field(
        default=None,
        max_length=1000,
        description="Natural language input: 'dal chawal medium bowl'",
    )

    # Manual mode: user provides exact values
    food_name: str | None = Field(default=None, max_length=255)
    brand_name: str | None = Field(default=None, max_length=255)
    portion_description: str | None = Field(default=None, max_length=255)
    portion_grams: float | None = Field(default=None, ge=0.0, le=10000.0)
    calories: float | None = Field(default=None, ge=0.0, le=10000.0)
    protein_g: float | None = Field(default=None, ge=0.0, le=1000.0)
    carbs_g: float | None = Field(default=None, ge=0.0, le=2000.0)
    fat_g: float | None = Field(default=None, ge=0.0, le=500.0)
    fiber_g: float | None = Field(default=None, ge=0.0, le=200.0)

    meal_type: MealType = MealType.SNACK
    logged_at: datetime | None = Field(
        default=None,
        description="When the food was actually eaten. Defaults to now.",
    )

    @model_validator(mode="after")
    def validate_input_mode(self) -> "FoodLogCreate":
        has_raw = self.raw_input is not None and len(self.raw_input.strip()) > 0
        has_manual = self.food_name is not None and self.calories is not None

        if not has_raw and not has_manual:
            raise ValueError(
                "Provide either raw_input (quick log) or "
                "food_name + calories (manual log)"
            )
        return self

    @property
    def is_quick_log(self) -> bool:
        return self.raw_input is not None and len(self.raw_input.strip()) > 0

    @property
    def is_manual_log(self) -> bool:
        return self.food_name is not None and self.calories is not None


class FoodLogUpdate(TrackBaseSchema):
    """
    Body for PUT /api/v1/food-logs/{log_id}

    Used for user corrections. When calories or portion are changed,
    a CorrectionEvent is automatically recorded.
    """
    food_name: str | None = Field(default=None, max_length=255)
    brand_name: str | None = Field(default=None, max_length=255)
    portion_description: str | None = Field(default=None, max_length=255)
    portion_grams: float | None = Field(default=None, ge=0.0, le=10000.0)
    calories: float | None = Field(default=None, ge=0.0, le=10000.0)
    protein_g: float | None = Field(default=None, ge=0.0, le=1000.0)
    carbs_g: float | None = Field(default=None, ge=0.0, le=2000.0)
    fat_g: float | None = Field(default=None, ge=0.0, le=500.0)
    fiber_g: float | None = Field(default=None, ge=0.0, le=200.0)
    meal_type: MealType | None = None
    logged_at: datetime | None = None


# ── Response Schemas ───────────────────────────────────────────────────────────

class FoodLogResponse(TrackBaseSchema):
    """Full food log response."""
    id: UUID
    user_id: UUID
    logged_at: datetime
    created_at: datetime
    updated_at: datetime

    raw_input: str | None
    meal_type: str
    food_name: str
    brand_name: str | None
    portion_description: str | None
    portion_grams: float | None

    calories: float
    protein_g: float | None
    carbs_g: float | None
    fat_g: float | None
    fiber_g: float | None

    estimation_source: str
    confidence_score: float
    confidence_level: str
    assumptions: list[str]

    is_corrected: bool
    original_calories: float | None

    image_url: str | None


class FoodLogSummary(TrackBaseSchema):
    """Lightweight food log for list views."""
    id: UUID
    logged_at: datetime
    meal_type: str
    food_name: str
    calories: float
    confidence_level: str
    estimation_source: str
    is_corrected: bool


class DailyFoodSummary(TrackBaseSchema):
    """Aggregated daily nutrition from food logs."""
    date: str  # ISO date string "YYYY-MM-DD"
    total_calories: float
    total_protein_g: float
    total_carbs_g: float
    total_fat_g: float
    food_count: int
    logs: list[FoodLogSummary]

File: mobile/lib/core/models/food_log.dart
class FoodLog {
  final String id;
  final String userId;
  final String foodName;
  final String? brandName;
  final String? rawInput;
  final MealType mealType;
  final String? portionDescription;
  final double? portionGrams;
  final double calories;
  final double? proteinG;
  final double? carbsG;
  final double? fatG;
  final double? fiberG;
  final EstimationSource estimationSource;
  final double confidenceScore;
  final ConfidenceLevel confidenceLevel;
  final List<String> assumptions;
  final bool isCorrected;
  final DateTime loggedAt;
  final DateTime createdAt;

  const FoodLog({
    required this.id,
    required this.userId,
    required this.foodName,
    this.brandName,
    this.rawInput,
    required this.mealType,
    this.portionDescription,
    this.portionGrams,
    required this.calories,
    this.proteinG,
    this.carbsG,
    this.fatG,
    this.fiberG,
    required this.estimationSource,
    required this.confidenceScore,
    required this.confidenceLevel,
    required this.assumptions,
    required this.isCorrected,
    required this.loggedAt,
    required this.createdAt,
  });

  factory FoodLog.fromJson(Map<String, dynamic> json) => FoodLog(
        id: json['id'] as String,
        userId: json['user_id'] as String,
        foodName: json['food_name'] as String,
        brandName: json['brand_name'] as String?,
        rawInput: json['raw_input'] as String?,
        mealType: MealType.fromString(json['meal_type'] as String),
        portionDescription: json['portion_description'] as String?,
        portionGrams: (json['portion_grams'] as num?)?.toDouble(),
        calories: (json['calories'] as num).toDouble(),
        proteinG: (json['protein_g'] as num?)?.toDouble(),
        carbsG: (json['carbs_g'] as num?)?.toDouble(),
        fatG: (json['fat_g'] as num?)?.toDouble(),
        fiberG: (json['fiber_g'] as num?)?.toDouble(),
        estimationSource: EstimationSource.fromString(json['estimation_source'] as String),
        confidenceScore: (json['confidence_score'] as num).toDouble(),
        confidenceLevel: ConfidenceLevel.fromString(json['confidence_level'] as String),
        assumptions: (json['assumptions'] as List<dynamic>).cast<String>(),
        isCorrected: json['is_corrected'] as bool,
        loggedAt: DateTime.parse(json['logged_at'] as String),
        createdAt: DateTime.parse(json['created_at'] as String),
      );
}

class DailyFoodSummary {
  final String date;
  final double totalCalories;
  final double totalProteinG;
  final double totalCarbsG;
  final double totalFatG;
  final int foodCount;
  final List<FoodLog> logs;

  const DailyFoodSummary({
    required this.date,
    required this.totalCalories,
    required this.totalProteinG,
    required this.totalCarbsG,
    required this.totalFatG,
    required this.foodCount,
    required this.logs,
  });

  factory DailyFoodSummary.fromJson(Map<String, dynamic> json) => DailyFoodSummary(
        date: json['date'] as String,
        totalCalories: (json['total_calories'] as num).toDouble(),
        totalProteinG: (json['total_protein_g'] as num).toDouble(),
        totalCarbsG: (json['total_carbs_g'] as num).toDouble(),
        totalFatG: (json['total_fat_g'] as num).toDouble(),
        foodCount: json['food_count'] as int,
        logs: (json['logs'] as List<dynamic>)
            .map((e) => FoodLog.fromJson(e as Map<String, dynamic>))
            .toList(),
      );
}

enum MealType {
  breakfast('breakfast', 'Breakfast'),
  lunch('lunch', 'Lunch'),
  dinner('dinner', 'Dinner'),
  snack('snack', 'Snack'),
  preWorkout('pre_workout', 'Pre-Workout'),
  postWorkout('post_workout', 'Post-Workout');

  const MealType(this.value, this.label);
  final String value;
  final String label;

  static MealType fromString(String s) =>
      MealType.values.firstWhere((e) => e.value == s, orElse: () => MealType.snack);
}

enum EstimationSource {
  memory('memory'),
  dataset('dataset'),
  llm('llm'),
  manual('manual'),
  photo('photo'),
  healthConnect('health_connect');

  const EstimationSource(this.value);
  final String value;

  static EstimationSource fromString(String s) =>
      EstimationSource.values.firstWhere((e) => e.value == s, orElse: () => EstimationSource.llm);
}

enum ConfidenceLevel {
  confirmed('confirmed'),
  estimated('estimated'),
  uncertain('uncertain');

  const ConfidenceLevel(this.value);
  final String value;

  static ConfidenceLevel fromString(String s) =>
      ConfidenceLevel.values.firstWhere((e) => e.value == s, orElse: () => ConfidenceLevel.estimated);
}
